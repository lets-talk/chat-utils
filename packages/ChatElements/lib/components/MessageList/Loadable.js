import React from 'react';
import path from 'path';
import Loadable from 'react-loadable';
import LoadingComponent from '../Loading';

const LoadableMessageList = Loadable({
  loader: () => import('./index'),
  loading: LoadingComponent,
  serverSideRequirePath: path.resolve(__dirname, './index'),
});

const MessageList = (props) => (<LoadableMessageList {...props} />);

export default MessageList;
