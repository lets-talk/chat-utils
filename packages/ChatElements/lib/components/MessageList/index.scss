@import "./config.scss";

.letstalk-messages-container {
  font-size: $font-size-normal;
  background-color: $widgetBox-MessageList-background;
  height: 50vh;
  min-height: 200px;
  overflow-y: auto;
  padding: 10px;
}
