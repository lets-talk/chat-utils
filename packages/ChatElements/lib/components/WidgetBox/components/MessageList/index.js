import React, { Component } from 'react';
import PropTypes from 'prop-types';
// import ImmutablePropTypes from 'react-immutable-proptypes';
import MessageBox from '../../../MessageBox';
import './index.scss';

const scrollToBottom = () => {
  const messagesDiv = document.getElementById('messages');
  messagesDiv.scrollTop = messagesDiv.scrollHeight;
};

class MessageList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      scrollBottom: 0,
    };
  }

  componentWillReceiveProps() {
    if (!this.mlistRef) { return; }
    this.setState({
      scrollBottom: this.getBottom(this.mlistRef),
    });
  }

  componentDidUpdate() {
    const e = this.mlistRef;
    if (!e) { return; }

    const bottom = this.getBottom(e);
    if (this.props.toBottomHeight === '100%' || bottom < this.props.toBottomHeight) {
      // scroll to bottom
      e.scrollTop = e.scrollHeight;
    } else if (this.props.lockable === true) {
      e.scrollTop = e.scrollHeight - e.offsetHeight - this.state.scrollBottom;
    }
  }

  onOpen(item, i, e) {
    if (this.props.onOpen instanceof Function) { this.props.onOpen(item, i, e); }
  }

  onDownload(item, i, e) {
    if (this.props.onDownload instanceof Function) { this.props.onDownload(item, i, e); }
  }

  onClick(item, i, e) {
    if (this.props.onClick instanceof Function) { this.props.onClick(item, i, e); }
  }

  onTitleClick(item, i, e) {
    if (this.props.onTitleClick instanceof Function) { this.props.onTitleClick(item, i, e); }
  }

  onForwardClick(item, i, e) {
    if (this.props.onForwardClick instanceof Function) { this.props.onForwardClick(item, i, e); }
  }

  getBottom(e) {
    return e.scrollHeight - e.scrollTop - e.offsetHeight;
  }

  // componentDidMount() {
  //   scrollToBottom();
  // }
  //
  // componentDidUpdate() {
  //   scrollToBottom();
  // }

  render() {
    return (
      <div id="messages" className="letstalk-messages-container">
        {
          this.props.messages.map((message, index) =>
            (
              <div className="letstalk-message" key={index}>
                <MessageBox
                  key={`message-${index}`}
                  {...message}
                  onOpen={this.props.onOpen && ((e) => this.onOpen(message, index, e))}
                  onDownload={this.props.onDownload && ((e) => this.onDownload(message, index, e))}
                  onTitleClick={this.props.onDownload && ((e) => this.onTitleClick(message, index, e))}
                  onForwardClick={this.props.onForwardClick && ((e) => this.onForwardClick(message, index, e))}
                  onClick={this.props.onClick && ((e) => this.onClick(message, index, e))}
                />
              </div>
            ))
        }
      </div>
    );
  }
}

MessageList.propTypes = {
  messages: PropTypes.array,
  onClick: PropTypes.func,
  onTitleClick: PropTypes.func,
  onForwardClick: PropTypes.func,
  onOpen: PropTypes.func,
  onDownload: PropTypes.func,
  lockable: PropTypes.bool,
  toBottomHeight: PropTypes.number,
};

MessageList.defaultProps = {
  messages: [],
  onClick: null,
  onTitleClick: null,
  onForwardClick: null,
  onOpen: null,
  onDownload: null,
  lockable: false,
  toBottomHeight: 300,
};

export default MessageList;
