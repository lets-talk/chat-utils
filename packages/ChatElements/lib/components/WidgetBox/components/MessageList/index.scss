@import "../common.scss";

.letstalk-messages-container {
  background-color: var(--color-white);
  height: 50vh;
  min-height: 200px;
  max-height: 410px;
  overflow-y: scroll;
  padding: 10px;
}

.full-screen {
  .letstalk-messages-container {
    @include letstalk-messages-container;
  }
}

@media screen and (max-width: 800px)  {
  .letstalk-messages-container {
    @include letstalk-messages-container;
  }
}
