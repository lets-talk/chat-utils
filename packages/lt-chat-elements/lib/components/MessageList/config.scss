@import '../../common/styles/variables';

$widgetBox-MessageList-background: $color-white;
