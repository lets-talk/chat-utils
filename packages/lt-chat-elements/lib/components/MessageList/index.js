import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { withAutoScroll } from '../../utils/hoc';

import MessageBox from '../MessageBox';
import './index.scss';

class MessageList extends Component {
  onOpen(item, i, e) {
    if (this.props.onOpen instanceof Function) { this.props.onOpen(item, i, e); }
  }

  onDownload(item, i, e) {
    if (this.props.onDownload instanceof Function) { this.props.onDownload(item, i, e); }
  }

  onClick(item, i, e) {
    if (this.props.onClick instanceof Function) { this.props.onClick(item, i, e); }
  }

  onTitleClick(item, i, e) {
    if (this.props.onTitleClick instanceof Function) { this.props.onTitleClick(item, i, e); }
  }

  onForwardClick(item, i, e) {
    if (this.props.onForwardClick instanceof Function) { this.props.onForwardClick(item, i, e); }
  }

  render() {
    return (
      <div ref={this.props.cmpRef} id="messages" className="letstalk-messages-container">
        {
          this.props.messages.map((message, index) =>
            (
              <div className="letstalk-message" key={index}>
                <MessageBox
                  key={`message-${index}`}
                  {...message}
                  onOpen={this.props.onOpen && ((e) => this.onOpen(message, index, e))}
                  onDownload={this.props.onDownload && ((e) => this.onDownload(message, index, e))}
                  onTitleClick={this.props.onDownload && ((e) => this.onTitleClick(message, index, e))}
                  onForwardClick={this.props.onForwardClick && ((e) => this.onForwardClick(message, index, e))}
                  onClick={this.props.onClick && ((e) => this.onClick(message, index, e))}
                  onActionMessageClick={this.props.onActionMessageClick}
                />
              </div>
            ))
        }
      </div>
    );
  }
}

MessageList.propTypes = {
  messages: PropTypes.array,
  onClick: PropTypes.func,
  onTitleClick: PropTypes.func,
  onForwardClick: PropTypes.func,
  onOpen: PropTypes.func,
  onDownload: PropTypes.func,
  onActionMessageClick: PropTypes.func,
  cmpRef: PropTypes.func,
};

MessageList.defaultProps = {
  messages: [],
  onClick: null,
  onTitleClick: null,
  onForwardClick: null,
  onOpen: null,
  onDownload: null,
  onActionMessageClick: null,
  cmpRef: null,
};

const autoScrollOptions = { threshold: 300, direction: 'bottom' };
export default withAutoScroll(autoScrollOptions)(MessageList);
