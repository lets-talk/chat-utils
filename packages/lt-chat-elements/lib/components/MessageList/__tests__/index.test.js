import React from 'react';
import { shallow } from 'enzyme';

import MessageList from '../index';

describe('<MessageList /> component', () => {
  it('matches snapshot', () => {
    const list = shallow(<MessageList><div>Child 1</div><div>Child 2</div></MessageList>);
    expect(list).toMatchSnapshot();
  });

  it('renders children inside', () => {
    const list = shallow(<MessageList><div>Child 1</div><div>Child 2</div></MessageList>);
    expect(list.find('div').length).toBe(2);
  });
});
