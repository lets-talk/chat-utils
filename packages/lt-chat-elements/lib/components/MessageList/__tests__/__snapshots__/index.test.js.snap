// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageList /> component matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <_class2>
    <div>
      Child 1
    </div>
    <div>
      Child 2
    </div>
  </_class2>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div>
          Child 1
        </div>,
        <div>
          Child 2
        </div>,
      ],
      "className": "LT-MessageList-Container",
      "cmpRef": [Function],
      "onClick": null,
      "onDownload": null,
      "onForwardClick": null,
      "onOpen": null,
      "onTitleClick": null,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Child 1",
        },
        "ref": null,
        "rendered": "Child 1",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Child 2",
        },
        "ref": null,
        "rendered": "Child 2",
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div>
            Child 1
          </div>,
          <div>
            Child 2
          </div>,
        ],
        "className": "LT-MessageList-Container",
        "cmpRef": [Function],
        "onClick": null,
        "onDownload": null,
        "onForwardClick": null,
        "onOpen": null,
        "onTitleClick": null,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Child 1",
          },
          "ref": null,
          "rendered": "Child 1",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Child 2",
          },
          "ref": null,
          "rendered": "Child 2",
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
